src/fitnesse/testrunner/run/PartitioningTestRunFactory.java

```diff
@@ -0,0 +1,103 @@
+  public static final String PARTITION_COUNT_ARG = "partitionCount";
+  public static final String PARTITION_INDEX_ARG = "partitionIndex";
+
+  private Function<List<WikiPage>, ? extends PagesByTestSystem> factory;
+  private Function<List<WikiPage>, PageListPartitioner> partitioner;
+
+  public PartitioningTestRunFactory() {
+    this(PagesByTestSystem::new,
+      pages -> new PageListPartitionerImpl(new EqualLengthListPartitioner<>()));
+  }
+
+  public PartitioningTestRunFactory(Function<List<WikiPage>, ? extends PagesByTestSystem> factory,
+                                    Function<List<WikiPage>, PageListPartitioner> partitioner) {
+    this.factory = factory;
+    this.partitioner = partitioner;
+  }
+
+  @Override
+  public boolean canRun(List<WikiPage> pages) {
+    return true;
+  }
+
+  public TestRun createRun(List<WikiPage> pages) {
+    PagesByTestSystem pagesByTestSystem = getPagesByTestSystem(pages);
+    return createRun(pagesByTestSystem);
+  }
+
+  public PagePositions findPagePositions(List<WikiPage> pages) {
+    if (pages.isEmpty()) {
+      return new PagePositions();
+    }
+    WikiPage page = pages.get(0);
+    PageListPartitioner pageListPartitioner = partitioner.apply(pages);
+    int partitionCount = getPartitionCount(page);
+    return pageListPartitioner.findPagePositions(factory, pages, partitionCount);
+  }
+
+  protected PagesByTestSystem getPagesByTestSystem(List<WikiPage> pages) {
+    if (pages.isEmpty()) {
+      return factory.apply(pages);
+    }
+    WikiPage page = pages.get(0);
+    PageListPartitioner pageListPartitioner = partitioner.apply(pages);
+    int partitionCount = getPartitionCount(page);
+    if (partitionCount == 1) {
+      return factory.apply(pages);
+    } else {
+      int partitionIndex = getPartitionIndex(page);
+      if (partitionIndex >= partitionCount) {
+        throw new IllegalArgumentException(PARTITION_COUNT_ARG + " must be larger than " + PARTITION_INDEX_ARG);
+      }
+      return pageListPartitioner.partition(factory, pages, partitionCount, partitionIndex);
+    }
+  }
+
+  protected TestRun createRun(PagesByTestSystem pagesByTestSystem) {
+    return new PerTestSystemTestRun(pagesByTestSystem);
+  }
+
+  protected int getPartitionCount(WikiPage page) {
+    int partitionCount = 1;
+    String partitionCountStr = page.getVariable(PARTITION_COUNT_ARG);
+    if (StringUtils.isNotEmpty(partitionCountStr)) {
+      partitionCount = Integer.parseInt(partitionCountStr);
+    }
+    return partitionCount;
+  }
+
+  protected int getPartitionIndex(WikiPage page) {
+    int partitionIndex = 0;
+    String partitionIndexStr = page.getVariable(PARTITION_INDEX_ARG);
+    if (StringUtils.isNotEmpty(partitionIndexStr)) {
+      partitionIndex = Integer.parseInt(partitionIndexStr);
+    }
+    return partitionIndex;
+  }
+
+  public Function<List<WikiPage>, ? extends PagesByTestSystem> getFactory() {
+    return factory;
```

Comment:
Remove usage of generic wildcard type and instead use a bounded parameter.

Reply:
This is a bounded parameter, isn't it: the function may return any subclass of `PagesByTestSystem`. I don't want to force that the function returns `PagesByTestSystem` if any subclass suffices.

If I misunderstand you: can you make an explicit suggestion what to change this to?

Reply:
> 
> 
> This is a bounded parameter, isn't it: the function may return any subclass of `PagesByTestSystem`. I don't want to force that the function returns `PagesByTestSystem` if any subclass suffices.
> 
> If I misunderstand you: can you make an explicit suggestion what to change this to?

No it isn't the parameter can be of any subclass, because of the wildcard type used as return type. To clarify what I mean I will share a screenshot:
![image](https://user-images.githubusercontent.com/4233200/75722418-75f77900-5cda-11ea-955c-18fb6cb6baaa.png)


In your case I would change it to: `Function<List<WikiPage>, PagesByTestSystem>`

Reply:
Sorry I'm stopping this discussion now. I'm not changing it. A `Function<A,B>` indicates a function that will return an instance of type `B` this is fundamentally different than the type parameter of a List. 

This signature says: you get a function that takes a list WikiPages and returns a PagesByTestSystem or a subclass of PagesByTestSystem. 
(The type parameter of a List also says something about its input parameters, for instance the `add` method will accept only instances of the type supplied, so you don't know what you can add to a `List<? extends A>`. This is not the case for the Function interface.)

The consumer of this function is not any client that now has to deal with a function with a wildcard. The consumer of this function in the partitioning algorithm and it has a wildcard to allow customisation by users. So other functions can be used/supplied.